atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketAssigner.java

```diff
@@ -0,0 +1,216 @@
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+
+public final class DefaultBucketAssigner {
+    private static final SafeLogger log = SafeLoggerFactory.get(DefaultBucketAssigner.class);
+    private static final int MAX_BUCKETS_PER_ITERATION = 5;
+
+    private final SweepBucketAssignerStateMachineTable sweepBucketAssignerStateMachineTable;
+    private final PuncherStore puncherStore;
+
+    private final BucketAssignerMetrics metrics;
+    private final BucketWriter bucketWriter;
+    private final Clock clock;
+
+    @VisibleForTesting
+    DefaultBucketAssigner(
+            SweepBucketAssignerStateMachineTable sweepBucketAssignerStateMachineTable,
+            PuncherStore puncherStore,
+            BucketWriter bucketWriter,
+            BucketAssignerMetrics metrics,
+            Clock clock) {
+        this.sweepBucketAssignerStateMachineTable = sweepBucketAssignerStateMachineTable;
+        this.puncherStore = puncherStore;
+        this.metrics = metrics;
+        this.bucketWriter = bucketWriter;
+        this.clock = clock;
+    }
+
+    public static DefaultBucketAssigner create(
+            SweepBucketAssignerStateMachineTable sweepBucketAssignerStateMachineTable,
+            PuncherStore puncherStore,
+            BucketWriter bucketWriter,
+            BucketAssignerMetrics metrics) {
+        return new DefaultBucketAssigner(
+                sweepBucketAssignerStateMachineTable, puncherStore, bucketWriter, metrics, Clock.systemUTC());
+    }
+
+    public void run() {
+        // TODO(mdaudali): Take a lock, or use Lockable<ExclusiveTask> to externalise the locking.
+        //  then, this may end up implementing Runnable.
+        BucketState bucketState = createNewBucket();
+
+        // We cap the number of buckets we create per iteration to avoid a single thread getting stuck creating
+        // buckets. This is a safety measure in case there's a bug in the bucket creation logic that causes it to
+        // infinitely create the same bucket or return CLOSED over and over again.
+        for (int i = 1; i < MAX_BUCKETS_PER_ITERATION && bucketState == BucketState.CLOSED; i++) {
+            bucketState = createNewBucket();
+            if (i == MAX_BUCKETS_PER_ITERATION - 1 && bucketState == BucketState.CLOSED) {
+                metrics.createdMaxBucketsInSingleIteration();
+                log.warn("Reached the maximum number of buckets created per iteration. This is most likely because"
+                        + " we're very far behind, or because there's a bug in the bucket creation logic that"
+                        + " causes it to infinitely create the same bucket or return CLOSED over and over again.");
+            }
+        }
+    }
+
+    private BucketState createNewBucket() {
```

Comment:
yeah maybe this should return BSAI (even if the underlying state machine just uses BS)

Reply:
(There's perhaps too many bucket states, but this is separate from the bucket assigner state - this was the BucketState that represents what happened on this iteration, for use in a metric / auto scaling. It might end up being superfluous)